### ISSUE TITLE
[java] False positive MissingStaticMethodInNonInstantiatableClass when inheritors are instantiable

### ISSUE THREAD

**Issue by skozlov** [pmd/pmd#2102]
**Affects PMD Version:** 6.19.0

**Rule:** [MissingStaticMethodInNonInstantiatableClass](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#missingstaticmethodinnoninstantiatableclass)

**Description:** if a class has only private constructors, it still may contain inner static inheritors. It is useful when implementing an [ADT](https://en.wikipedia.org/wiki/Algebraic_data_type).

**Code Sample demonstrating the issue:**

```java
public abstract class MyADT {
    private MyADT() {
    }

    public abstract <R> R map(
            Function<String, ? extends R> onString,
            Function<Integer, ? extends R> onInt
    );

    public static final class StringHolder extends MyADT {
        private final String string;

        public StringHolder(String string) {
            this.string = string;
        }

        @Override
        public <R> R map(Function<String, ? extends R> onString, Function<Integer, ? extends R> onInt) {
            return onString.apply(string);
        }
    }

    public static final class IntHolder extends MyADT {
        private final Integer integer;

        public IntHolder(Integer integer) {
            this.integer = integer;
        }

        @Override
        public <R> R map(Function<String, ? extends R> onString, Function<Integer, ? extends R> onInt) {
            return onInt.apply(integer);
        }
    }
}
```

**Running PMD through:** *Gradle*
